{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "8e2149fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "import pandas as pd\n",
    "import xarray as xr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "d03062ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "chrom=22"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "74a24382",
   "metadata": {},
   "outputs": [],
   "source": [
    "# open anndata \n",
    "my_file = \"/share/ScratchGeneral/anncuo/OneK1K/expression_objects/sce\"+str(chrom)+\".h5ad\"\n",
    "adata = sc.read(my_file)\n",
    "# sparse to dense\n",
    "mat = adata.raw.X.todense()\n",
    "# make pandas dataframe\n",
    "mat_df = pd.DataFrame(data=mat.T, index=adata.raw.var.index, columns=adata.obs.index)\n",
    "# turn into xr array\n",
    "phenotype = xr.DataArray(mat_df.values, dims=[\"trait\", \"cell\"], coords={\"trait\": mat_df.index.values, \"cell\": mat_df.columns.values})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "3bd705df",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><svg style=\"position: absolute; width: 0; height: 0; overflow: hidden\">\n",
       "<defs>\n",
       "<symbol id=\"icon-database\" viewBox=\"0 0 32 32\">\n",
       "<path d=\"M16 0c-8.837 0-16 2.239-16 5v4c0 2.761 7.163 5 16 5s16-2.239 16-5v-4c0-2.761-7.163-5-16-5z\"></path>\n",
       "<path d=\"M16 17c-8.837 0-16-2.239-16-5v6c0 2.761 7.163 5 16 5s16-2.239 16-5v-6c0 2.761-7.163 5-16 5z\"></path>\n",
       "<path d=\"M16 26c-8.837 0-16-2.239-16-5v6c0 2.761 7.163 5 16 5s16-2.239 16-5v-6c0 2.761-7.163 5-16 5z\"></path>\n",
       "</symbol>\n",
       "<symbol id=\"icon-file-text2\" viewBox=\"0 0 32 32\">\n",
       "<path d=\"M28.681 7.159c-0.694-0.947-1.662-2.053-2.724-3.116s-2.169-2.030-3.116-2.724c-1.612-1.182-2.393-1.319-2.841-1.319h-15.5c-1.378 0-2.5 1.121-2.5 2.5v27c0 1.378 1.122 2.5 2.5 2.5h23c1.378 0 2.5-1.122 2.5-2.5v-19.5c0-0.448-0.137-1.23-1.319-2.841zM24.543 5.457c0.959 0.959 1.712 1.825 2.268 2.543h-4.811v-4.811c0.718 0.556 1.584 1.309 2.543 2.268zM28 29.5c0 0.271-0.229 0.5-0.5 0.5h-23c-0.271 0-0.5-0.229-0.5-0.5v-27c0-0.271 0.229-0.5 0.5-0.5 0 0 15.499-0 15.5 0v7c0 0.552 0.448 1 1 1h7v19.5z\"></path>\n",
       "<path d=\"M23 26h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z\"></path>\n",
       "<path d=\"M23 22h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z\"></path>\n",
       "<path d=\"M23 18h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z\"></path>\n",
       "</symbol>\n",
       "</defs>\n",
       "</svg>\n",
       "<style>/* CSS stylesheet for displaying xarray objects in jupyterlab.\n",
       " *\n",
       " */\n",
       "\n",
       ":root {\n",
       "  --xr-font-color0: var(--jp-content-font-color0, rgba(0, 0, 0, 1));\n",
       "  --xr-font-color2: var(--jp-content-font-color2, rgba(0, 0, 0, 0.54));\n",
       "  --xr-font-color3: var(--jp-content-font-color3, rgba(0, 0, 0, 0.38));\n",
       "  --xr-border-color: var(--jp-border-color2, #e0e0e0);\n",
       "  --xr-disabled-color: var(--jp-layout-color3, #bdbdbd);\n",
       "  --xr-background-color: var(--jp-layout-color0, white);\n",
       "  --xr-background-color-row-even: var(--jp-layout-color1, white);\n",
       "  --xr-background-color-row-odd: var(--jp-layout-color2, #eeeeee);\n",
       "}\n",
       "\n",
       "html[theme=dark],\n",
       "body.vscode-dark {\n",
       "  --xr-font-color0: rgba(255, 255, 255, 1);\n",
       "  --xr-font-color2: rgba(255, 255, 255, 0.54);\n",
       "  --xr-font-color3: rgba(255, 255, 255, 0.38);\n",
       "  --xr-border-color: #1F1F1F;\n",
       "  --xr-disabled-color: #515151;\n",
       "  --xr-background-color: #111111;\n",
       "  --xr-background-color-row-even: #111111;\n",
       "  --xr-background-color-row-odd: #313131;\n",
       "}\n",
       "\n",
       ".xr-wrap {\n",
       "  display: block !important;\n",
       "  min-width: 300px;\n",
       "  max-width: 700px;\n",
       "}\n",
       "\n",
       ".xr-text-repr-fallback {\n",
       "  /* fallback to plain text repr when CSS is not injected (untrusted notebook) */\n",
       "  display: none;\n",
       "}\n",
       "\n",
       ".xr-header {\n",
       "  padding-top: 6px;\n",
       "  padding-bottom: 6px;\n",
       "  margin-bottom: 4px;\n",
       "  border-bottom: solid 1px var(--xr-border-color);\n",
       "}\n",
       "\n",
       ".xr-header > div,\n",
       ".xr-header > ul {\n",
       "  display: inline;\n",
       "  margin-top: 0;\n",
       "  margin-bottom: 0;\n",
       "}\n",
       "\n",
       ".xr-obj-type,\n",
       ".xr-array-name {\n",
       "  margin-left: 2px;\n",
       "  margin-right: 10px;\n",
       "}\n",
       "\n",
       ".xr-obj-type {\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-sections {\n",
       "  padding-left: 0 !important;\n",
       "  display: grid;\n",
       "  grid-template-columns: 150px auto auto 1fr 20px 20px;\n",
       "}\n",
       "\n",
       ".xr-section-item {\n",
       "  display: contents;\n",
       "}\n",
       "\n",
       ".xr-section-item input {\n",
       "  display: none;\n",
       "}\n",
       "\n",
       ".xr-section-item input + label {\n",
       "  color: var(--xr-disabled-color);\n",
       "}\n",
       "\n",
       ".xr-section-item input:enabled + label {\n",
       "  cursor: pointer;\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-section-item input:enabled + label:hover {\n",
       "  color: var(--xr-font-color0);\n",
       "}\n",
       "\n",
       ".xr-section-summary {\n",
       "  grid-column: 1;\n",
       "  color: var(--xr-font-color2);\n",
       "  font-weight: 500;\n",
       "}\n",
       "\n",
       ".xr-section-summary > span {\n",
       "  display: inline-block;\n",
       "  padding-left: 0.5em;\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:disabled + label {\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-section-summary-in + label:before {\n",
       "  display: inline-block;\n",
       "  content: '►';\n",
       "  font-size: 11px;\n",
       "  width: 15px;\n",
       "  text-align: center;\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:disabled + label:before {\n",
       "  color: var(--xr-disabled-color);\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:checked + label:before {\n",
       "  content: '▼';\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:checked + label > span {\n",
       "  display: none;\n",
       "}\n",
       "\n",
       ".xr-section-summary,\n",
       ".xr-section-inline-details {\n",
       "  padding-top: 4px;\n",
       "  padding-bottom: 4px;\n",
       "}\n",
       "\n",
       ".xr-section-inline-details {\n",
       "  grid-column: 2 / -1;\n",
       "}\n",
       "\n",
       ".xr-section-details {\n",
       "  display: none;\n",
       "  grid-column: 1 / -1;\n",
       "  margin-bottom: 5px;\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:checked ~ .xr-section-details {\n",
       "  display: contents;\n",
       "}\n",
       "\n",
       ".xr-array-wrap {\n",
       "  grid-column: 1 / -1;\n",
       "  display: grid;\n",
       "  grid-template-columns: 20px auto;\n",
       "}\n",
       "\n",
       ".xr-array-wrap > label {\n",
       "  grid-column: 1;\n",
       "  vertical-align: top;\n",
       "}\n",
       "\n",
       ".xr-preview {\n",
       "  color: var(--xr-font-color3);\n",
       "}\n",
       "\n",
       ".xr-array-preview,\n",
       ".xr-array-data {\n",
       "  padding: 0 5px !important;\n",
       "  grid-column: 2;\n",
       "}\n",
       "\n",
       ".xr-array-data,\n",
       ".xr-array-in:checked ~ .xr-array-preview {\n",
       "  display: none;\n",
       "}\n",
       "\n",
       ".xr-array-in:checked ~ .xr-array-data,\n",
       ".xr-array-preview {\n",
       "  display: inline-block;\n",
       "}\n",
       "\n",
       ".xr-dim-list {\n",
       "  display: inline-block !important;\n",
       "  list-style: none;\n",
       "  padding: 0 !important;\n",
       "  margin: 0;\n",
       "}\n",
       "\n",
       ".xr-dim-list li {\n",
       "  display: inline-block;\n",
       "  padding: 0;\n",
       "  margin: 0;\n",
       "}\n",
       "\n",
       ".xr-dim-list:before {\n",
       "  content: '(';\n",
       "}\n",
       "\n",
       ".xr-dim-list:after {\n",
       "  content: ')';\n",
       "}\n",
       "\n",
       ".xr-dim-list li:not(:last-child):after {\n",
       "  content: ',';\n",
       "  padding-right: 5px;\n",
       "}\n",
       "\n",
       ".xr-has-index {\n",
       "  font-weight: bold;\n",
       "}\n",
       "\n",
       ".xr-var-list,\n",
       ".xr-var-item {\n",
       "  display: contents;\n",
       "}\n",
       "\n",
       ".xr-var-item > div,\n",
       ".xr-var-item label,\n",
       ".xr-var-item > .xr-var-name span {\n",
       "  background-color: var(--xr-background-color-row-even);\n",
       "  margin-bottom: 0;\n",
       "}\n",
       "\n",
       ".xr-var-item > .xr-var-name:hover span {\n",
       "  padding-right: 5px;\n",
       "}\n",
       "\n",
       ".xr-var-list > li:nth-child(odd) > div,\n",
       ".xr-var-list > li:nth-child(odd) > label,\n",
       ".xr-var-list > li:nth-child(odd) > .xr-var-name span {\n",
       "  background-color: var(--xr-background-color-row-odd);\n",
       "}\n",
       "\n",
       ".xr-var-name {\n",
       "  grid-column: 1;\n",
       "}\n",
       "\n",
       ".xr-var-dims {\n",
       "  grid-column: 2;\n",
       "}\n",
       "\n",
       ".xr-var-dtype {\n",
       "  grid-column: 3;\n",
       "  text-align: right;\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-var-preview {\n",
       "  grid-column: 4;\n",
       "}\n",
       "\n",
       ".xr-var-name,\n",
       ".xr-var-dims,\n",
       ".xr-var-dtype,\n",
       ".xr-preview,\n",
       ".xr-attrs dt {\n",
       "  white-space: nowrap;\n",
       "  overflow: hidden;\n",
       "  text-overflow: ellipsis;\n",
       "  padding-right: 10px;\n",
       "}\n",
       "\n",
       ".xr-var-name:hover,\n",
       ".xr-var-dims:hover,\n",
       ".xr-var-dtype:hover,\n",
       ".xr-attrs dt:hover {\n",
       "  overflow: visible;\n",
       "  width: auto;\n",
       "  z-index: 1;\n",
       "}\n",
       "\n",
       ".xr-var-attrs,\n",
       ".xr-var-data {\n",
       "  display: none;\n",
       "  background-color: var(--xr-background-color) !important;\n",
       "  padding-bottom: 5px !important;\n",
       "}\n",
       "\n",
       ".xr-var-attrs-in:checked ~ .xr-var-attrs,\n",
       ".xr-var-data-in:checked ~ .xr-var-data {\n",
       "  display: block;\n",
       "}\n",
       "\n",
       ".xr-var-data > table {\n",
       "  float: right;\n",
       "}\n",
       "\n",
       ".xr-var-name span,\n",
       ".xr-var-data,\n",
       ".xr-attrs {\n",
       "  padding-left: 25px !important;\n",
       "}\n",
       "\n",
       ".xr-attrs,\n",
       ".xr-var-attrs,\n",
       ".xr-var-data {\n",
       "  grid-column: 1 / -1;\n",
       "}\n",
       "\n",
       "dl.xr-attrs {\n",
       "  padding: 0;\n",
       "  margin: 0;\n",
       "  display: grid;\n",
       "  grid-template-columns: 125px auto;\n",
       "}\n",
       "\n",
       ".xr-attrs dt,\n",
       ".xr-attrs dd {\n",
       "  padding: 0;\n",
       "  margin: 0;\n",
       "  float: left;\n",
       "  padding-right: 10px;\n",
       "  width: auto;\n",
       "}\n",
       "\n",
       ".xr-attrs dt {\n",
       "  font-weight: normal;\n",
       "  grid-column: 1;\n",
       "}\n",
       "\n",
       ".xr-attrs dt:hover span {\n",
       "  display: inline-block;\n",
       "  background: var(--xr-background-color);\n",
       "  padding-right: 10px;\n",
       "}\n",
       "\n",
       ".xr-attrs dd {\n",
       "  grid-column: 2;\n",
       "  white-space: pre-wrap;\n",
       "  word-break: break-all;\n",
       "}\n",
       "\n",
       ".xr-icon-database,\n",
       ".xr-icon-file-text2 {\n",
       "  display: inline-block;\n",
       "  vertical-align: middle;\n",
       "  width: 1em;\n",
       "  height: 1.5em !important;\n",
       "  stroke-width: 0;\n",
       "  stroke: currentColor;\n",
       "  fill: currentColor;\n",
       "}\n",
       "</style><pre class='xr-text-repr-fallback'>&lt;xarray.DataArray (trait: 600, cell: 1272489)&gt;\n",
       "array([[0., 0., 0., ..., 0., 0., 0.],\n",
       "       [0., 0., 0., ..., 0., 0., 0.],\n",
       "       [0., 0., 0., ..., 0., 0., 0.],\n",
       "       ...,\n",
       "       [0., 0., 0., ..., 0., 0., 0.],\n",
       "       [0., 0., 0., ..., 0., 0., 0.],\n",
       "       [0., 0., 0., ..., 0., 0., 0.]])\n",
       "Coordinates:\n",
       "  * trait    (trait) object &#x27;AC002472.1&#x27; &#x27;KIAA1658&#x27; ... &#x27;AC000036.4&#x27;\n",
       "  * cell     (cell) object &#x27;AAACCTGAGAATGTTG-1&#x27; ... &#x27;TTTGTCATCTCGTATT-9&#x27;</pre><div class='xr-wrap' style='display:none'><div class='xr-header'><div class='xr-obj-type'>xarray.DataArray</div><div class='xr-array-name'></div><ul class='xr-dim-list'><li><span class='xr-has-index'>trait</span>: 600</li><li><span class='xr-has-index'>cell</span>: 1272489</li></ul></div><ul class='xr-sections'><li class='xr-section-item'><div class='xr-array-wrap'><input id='section-78651f77-3671-4b1a-911d-7c7bba527b73' class='xr-array-in' type='checkbox' checked><label for='section-78651f77-3671-4b1a-911d-7c7bba527b73' title='Show/hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-array-preview xr-preview'><span>0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 ... 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0</span></div><div class='xr-array-data'><pre>array([[0., 0., 0., ..., 0., 0., 0.],\n",
       "       [0., 0., 0., ..., 0., 0., 0.],\n",
       "       [0., 0., 0., ..., 0., 0., 0.],\n",
       "       ...,\n",
       "       [0., 0., 0., ..., 0., 0., 0.],\n",
       "       [0., 0., 0., ..., 0., 0., 0.],\n",
       "       [0., 0., 0., ..., 0., 0., 0.]])</pre></div></div></li><li class='xr-section-item'><input id='section-db234e00-8ddc-44c3-ba3f-2ae240c6f9d9' class='xr-section-summary-in' type='checkbox'  checked><label for='section-db234e00-8ddc-44c3-ba3f-2ae240c6f9d9' class='xr-section-summary' >Coordinates: <span>(2)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><ul class='xr-var-list'><li class='xr-var-item'><div class='xr-var-name'><span class='xr-has-index'>trait</span></div><div class='xr-var-dims'>(trait)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;AC002472.1&#x27; ... &#x27;AC000036.4&#x27;</div><input id='attrs-43861e2a-f114-4e3d-9ea3-41f02f9dfef8' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-43861e2a-f114-4e3d-9ea3-41f02f9dfef8' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-fc00b044-5090-417d-a32b-3ae70e202223' class='xr-var-data-in' type='checkbox'><label for='data-fc00b044-5090-417d-a32b-3ae70e202223' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;AC002472.1&#x27;, &#x27;KIAA1658&#x27;, &#x27;AC005003.1&#x27;, ..., &#x27;CHKB-AS1&#x27;, &#x27;CTA-384D8.33&#x27;,\n",
       "       &#x27;AC000036.4&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span class='xr-has-index'>cell</span></div><div class='xr-var-dims'>(cell)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;AAACCTGAGAATGTTG-1&#x27; ... &#x27;TTTGTC...</div><input id='attrs-04002cdc-a52d-4e2a-bde3-5e6586b455f0' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-04002cdc-a52d-4e2a-bde3-5e6586b455f0' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-05423609-0187-41f8-80c4-cf5ce37acd98' class='xr-var-data-in' type='checkbox'><label for='data-05423609-0187-41f8-80c4-cf5ce37acd98' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;AAACCTGAGAATGTTG-1&#x27;, &#x27;AAACCTGAGAGAACAG-1&#x27;, &#x27;AAACCTGAGCATGGCA-1&#x27;, ...,\n",
       "       &#x27;TTTGTCATCGCCGTGA-9&#x27;, &#x27;TTTGTCATCGCGGATC-9&#x27;, &#x27;TTTGTCATCTCGTATT-9&#x27;],\n",
       "      dtype=object)</pre></div></li></ul></div></li><li class='xr-section-item'><input id='section-4e8b4d1c-bcbc-470f-9879-fbe7bba4dcff' class='xr-section-summary-in' type='checkbox' disabled ><label for='section-4e8b4d1c-bcbc-470f-9879-fbe7bba4dcff' class='xr-section-summary'  title='Expand/collapse section'>Attributes: <span>(0)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><dl class='xr-attrs'></dl></div></li></ul></div></div>"
      ],
      "text/plain": [
       "<xarray.DataArray (trait: 600, cell: 1272489)>\n",
       "array([[0., 0., 0., ..., 0., 0., 0.],\n",
       "       [0., 0., 0., ..., 0., 0., 0.],\n",
       "       [0., 0., 0., ..., 0., 0., 0.],\n",
       "       ...,\n",
       "       [0., 0., 0., ..., 0., 0., 0.],\n",
       "       [0., 0., 0., ..., 0., 0., 0.],\n",
       "       [0., 0., 0., ..., 0., 0., 0.]])\n",
       "Coordinates:\n",
       "  * trait    (trait) object 'AC002472.1' 'KIAA1658' ... 'AC000036.4'\n",
       "  * cell     (cell) object 'AAACCTGAGAATGTTG-1' ... 'TTTGTCATCTCGTATT-9'"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "phenotype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "2fa87217",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>orig.ident</th>\n",
       "      <th>nCount_RNA</th>\n",
       "      <th>nFeature_RNA</th>\n",
       "      <th>pool</th>\n",
       "      <th>individual</th>\n",
       "      <th>percent.mt</th>\n",
       "      <th>latent</th>\n",
       "      <th>nCount_SCT</th>\n",
       "      <th>nFeature_SCT</th>\n",
       "      <th>cell_type</th>\n",
       "      <th>...</th>\n",
       "      <th>sex</th>\n",
       "      <th>predicted.celltype</th>\n",
       "      <th>predicted.celltype.l3</th>\n",
       "      <th>predicted.celltype.l4</th>\n",
       "      <th>predicted.celltype.l5</th>\n",
       "      <th>predicted.celltype.l6</th>\n",
       "      <th>G2M.Score</th>\n",
       "      <th>S.Score</th>\n",
       "      <th>Phase</th>\n",
       "      <th>barcode</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGAATGTTG-1</th>\n",
       "      <td>onek1k</td>\n",
       "      <td>4754</td>\n",
       "      <td>1201</td>\n",
       "      <td>pool_1</td>\n",
       "      <td>691_692</td>\n",
       "      <td>1.451409</td>\n",
       "      <td>b1</td>\n",
       "      <td>3226</td>\n",
       "      <td>1186</td>\n",
       "      <td>CD4+ KLRB1+ T cell</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>CD4+ KLRB1+ T cell</td>\n",
       "      <td>CD4 TEM</td>\n",
       "      <td>CD4 T</td>\n",
       "      <td>T cell</td>\n",
       "      <td>Lymphoid</td>\n",
       "      <td>-0.024395</td>\n",
       "      <td>0.008655</td>\n",
       "      <td>S</td>\n",
       "      <td>AAACCTGAGAATGTTG-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGAGAACAG-1</th>\n",
       "      <td>onek1k</td>\n",
       "      <td>2459</td>\n",
       "      <td>896</td>\n",
       "      <td>pool_1</td>\n",
       "      <td>693_694</td>\n",
       "      <td>3.700691</td>\n",
       "      <td>b1</td>\n",
       "      <td>2710</td>\n",
       "      <td>896</td>\n",
       "      <td>XCL1- NK</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>XCL1- NK</td>\n",
       "      <td>NK</td>\n",
       "      <td>NK cell</td>\n",
       "      <td>NK cell</td>\n",
       "      <td>Lymphoid</td>\n",
       "      <td>-0.026514</td>\n",
       "      <td>-0.015411</td>\n",
       "      <td>G1</td>\n",
       "      <td>AAACCTGAGAGAACAG-1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   orig.ident  nCount_RNA  nFeature_RNA    pool individual  \\\n",
       "AAACCTGAGAATGTTG-1     onek1k        4754          1201  pool_1    691_692   \n",
       "AAACCTGAGAGAACAG-1     onek1k        2459           896  pool_1    693_694   \n",
       "\n",
       "                    percent.mt latent  nCount_SCT  nFeature_SCT  \\\n",
       "AAACCTGAGAATGTTG-1    1.451409     b1        3226          1186   \n",
       "AAACCTGAGAGAACAG-1    3.700691     b1        2710           896   \n",
       "\n",
       "                             cell_type  ... sex  predicted.celltype  \\\n",
       "AAACCTGAGAATGTTG-1  CD4+ KLRB1+ T cell  ...   1  CD4+ KLRB1+ T cell   \n",
       "AAACCTGAGAGAACAG-1            XCL1- NK  ...   2            XCL1- NK   \n",
       "\n",
       "                    predicted.celltype.l3  predicted.celltype.l4  \\\n",
       "AAACCTGAGAATGTTG-1                CD4 TEM                  CD4 T   \n",
       "AAACCTGAGAGAACAG-1                     NK                NK cell   \n",
       "\n",
       "                   predicted.celltype.l5 predicted.celltype.l6 G2M.Score  \\\n",
       "AAACCTGAGAATGTTG-1                T cell              Lymphoid -0.024395   \n",
       "AAACCTGAGAGAACAG-1               NK cell              Lymphoid -0.026514   \n",
       "\n",
       "                     S.Score Phase             barcode  \n",
       "AAACCTGAGAATGTTG-1  0.008655     S  AAACCTGAGAATGTTG-1  \n",
       "AAACCTGAGAGAACAG-1 -0.015411    G1  AAACCTGAGAGAACAG-1  \n",
       "\n",
       "[2 rows x 23 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cov_filename = \"/share/ScratchGeneral/anncuo/OneK1K/covariates_new.csv\"\n",
    "cov_df = pd.read_csv(cov_filename, index_col=0)\n",
    "cov_df.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "9c2caaec",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['orig.ident', 'nCount_RNA', 'nFeature_RNA', 'pool', 'individual',\n",
       "       'percent.mt', 'latent', 'nCount_SCT', 'nFeature_SCT', 'cell_type',\n",
       "       'predicted.celltype.l2', 'predicted.celltype.l1', 'age', 'sex',\n",
       "       'predicted.celltype', 'predicted.celltype.l3', 'predicted.celltype.l4',\n",
       "       'predicted.celltype.l5', 'predicted.celltype.l6', 'G2M.Score',\n",
       "       'S.Score', 'Phase', 'barcode'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cov_df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "a4933dd7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['CD4 TEM', 'NK', 'CD4 Naive', 'CD8 TEM', 'B naive', 'CD8 TCM',\n",
       "       'CD4 TCM', 'B intermediate', nan, 'CD8 Naive', 'NK_CD56bright',\n",
       "       'B memory', 'CD16 Mono', 'MAIT', 'Treg', 'CD14 Mono', 'dnT',\n",
       "       'Eryth', 'NK Proliferating', 'CD4 CTL', 'HSPC', 'Platelet',\n",
       "       'Plasmablast', 'ILC', 'gdT', 'cDC2', 'pDC', 'ASDC',\n",
       "       'CD4 Proliferating', 'CD8 Proliferating', 'cDC1', 'Doublet'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cov_df['predicted.celltype.l2'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "f25d6539",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(71374, 23)"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#### select naive B cells\n",
    "cov_naive_Bcell = cov_df[cov_df['predicted.celltype.l2'] == 'B naive']\n",
    "cov_naive_Bcell.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "3f3b915f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>orig.ident</th>\n",
       "      <th>nCount_RNA</th>\n",
       "      <th>nFeature_RNA</th>\n",
       "      <th>pool</th>\n",
       "      <th>individual</th>\n",
       "      <th>percent.mt</th>\n",
       "      <th>latent</th>\n",
       "      <th>nCount_SCT</th>\n",
       "      <th>nFeature_SCT</th>\n",
       "      <th>cell_type</th>\n",
       "      <th>...</th>\n",
       "      <th>sex</th>\n",
       "      <th>predicted.celltype</th>\n",
       "      <th>predicted.celltype.l3</th>\n",
       "      <th>predicted.celltype.l4</th>\n",
       "      <th>predicted.celltype.l5</th>\n",
       "      <th>predicted.celltype.l6</th>\n",
       "      <th>G2M.Score</th>\n",
       "      <th>S.Score</th>\n",
       "      <th>Phase</th>\n",
       "      <th>barcode</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCTGCAACGATCT-1</th>\n",
       "      <td>onek1k</td>\n",
       "      <td>3847</td>\n",
       "      <td>884</td>\n",
       "      <td>pool_1</td>\n",
       "      <td>686_687</td>\n",
       "      <td>3.717182</td>\n",
       "      <td>b1</td>\n",
       "      <td>3085</td>\n",
       "      <td>883</td>\n",
       "      <td>TCL1A+ FCER2+ B cell</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>TCL1A+ FCER2+ B cell</td>\n",
       "      <td>B naive</td>\n",
       "      <td>B cell</td>\n",
       "      <td>B cell</td>\n",
       "      <td>Lymphoid</td>\n",
       "      <td>-0.026746</td>\n",
       "      <td>-0.000268</td>\n",
       "      <td>G1</td>\n",
       "      <td>AAACCTGCAACGATCT-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACGGGCAGTTCATG-1</th>\n",
       "      <td>onek1k</td>\n",
       "      <td>3092</td>\n",
       "      <td>811</td>\n",
       "      <td>pool_1</td>\n",
       "      <td>682_683</td>\n",
       "      <td>4.010349</td>\n",
       "      <td>b1</td>\n",
       "      <td>2941</td>\n",
       "      <td>811</td>\n",
       "      <td>TCL1A+ FCER2+ B cell</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>TCL1A+ FCER2+ B cell</td>\n",
       "      <td>B naive</td>\n",
       "      <td>B cell</td>\n",
       "      <td>B cell</td>\n",
       "      <td>Lymphoid</td>\n",
       "      <td>-0.022354</td>\n",
       "      <td>0.018777</td>\n",
       "      <td>S</td>\n",
       "      <td>AAACGGGCAGTTCATG-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACGGGGTCCCTACT-1</th>\n",
       "      <td>onek1k</td>\n",
       "      <td>3402</td>\n",
       "      <td>879</td>\n",
       "      <td>pool_1</td>\n",
       "      <td>692_693</td>\n",
       "      <td>2.704292</td>\n",
       "      <td>b1</td>\n",
       "      <td>3008</td>\n",
       "      <td>879</td>\n",
       "      <td>TCL1A+ FCER2+ B cell</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>TCL1A+ FCER2+ B cell</td>\n",
       "      <td>B naive</td>\n",
       "      <td>B cell</td>\n",
       "      <td>B cell</td>\n",
       "      <td>Lymphoid</td>\n",
       "      <td>0.013420</td>\n",
       "      <td>-0.004957</td>\n",
       "      <td>G2M</td>\n",
       "      <td>AAACGGGGTCCCTACT-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACGGGGTTGCGCAC-1</th>\n",
       "      <td>onek1k</td>\n",
       "      <td>2194</td>\n",
       "      <td>544</td>\n",
       "      <td>pool_1</td>\n",
       "      <td>683_684</td>\n",
       "      <td>3.783045</td>\n",
       "      <td>b1</td>\n",
       "      <td>2714</td>\n",
       "      <td>544</td>\n",
       "      <td>TCL1A+ FCER2+ B cell</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>TCL1A+ FCER2+ B cell</td>\n",
       "      <td>B naive</td>\n",
       "      <td>B cell</td>\n",
       "      <td>B cell</td>\n",
       "      <td>Lymphoid</td>\n",
       "      <td>0.006212</td>\n",
       "      <td>0.010032</td>\n",
       "      <td>S</td>\n",
       "      <td>AAACGGGGTTGCGCAC-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAAGATGGTTATGCGT-1</th>\n",
       "      <td>onek1k</td>\n",
       "      <td>1608</td>\n",
       "      <td>603</td>\n",
       "      <td>pool_1</td>\n",
       "      <td>686_687</td>\n",
       "      <td>4.415423</td>\n",
       "      <td>b1</td>\n",
       "      <td>2530</td>\n",
       "      <td>607</td>\n",
       "      <td>TCL1A+ FCER2+ B cell</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>TCL1A+ FCER2+ B cell</td>\n",
       "      <td>B naive</td>\n",
       "      <td>B cell</td>\n",
       "      <td>B cell</td>\n",
       "      <td>Lymphoid</td>\n",
       "      <td>-0.011067</td>\n",
       "      <td>0.004024</td>\n",
       "      <td>S</td>\n",
       "      <td>AAAGATGGTTATGCGT-1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   orig.ident  nCount_RNA  nFeature_RNA    pool individual  \\\n",
       "AAACCTGCAACGATCT-1     onek1k        3847           884  pool_1    686_687   \n",
       "AAACGGGCAGTTCATG-1     onek1k        3092           811  pool_1    682_683   \n",
       "AAACGGGGTCCCTACT-1     onek1k        3402           879  pool_1    692_693   \n",
       "AAACGGGGTTGCGCAC-1     onek1k        2194           544  pool_1    683_684   \n",
       "AAAGATGGTTATGCGT-1     onek1k        1608           603  pool_1    686_687   \n",
       "\n",
       "                    percent.mt latent  nCount_SCT  nFeature_SCT  \\\n",
       "AAACCTGCAACGATCT-1    3.717182     b1        3085           883   \n",
       "AAACGGGCAGTTCATG-1    4.010349     b1        2941           811   \n",
       "AAACGGGGTCCCTACT-1    2.704292     b1        3008           879   \n",
       "AAACGGGGTTGCGCAC-1    3.783045     b1        2714           544   \n",
       "AAAGATGGTTATGCGT-1    4.415423     b1        2530           607   \n",
       "\n",
       "                               cell_type  ... sex    predicted.celltype  \\\n",
       "AAACCTGCAACGATCT-1  TCL1A+ FCER2+ B cell  ...   1  TCL1A+ FCER2+ B cell   \n",
       "AAACGGGCAGTTCATG-1  TCL1A+ FCER2+ B cell  ...   1  TCL1A+ FCER2+ B cell   \n",
       "AAACGGGGTCCCTACT-1  TCL1A+ FCER2+ B cell  ...   2  TCL1A+ FCER2+ B cell   \n",
       "AAACGGGGTTGCGCAC-1  TCL1A+ FCER2+ B cell  ...   1  TCL1A+ FCER2+ B cell   \n",
       "AAAGATGGTTATGCGT-1  TCL1A+ FCER2+ B cell  ...   1  TCL1A+ FCER2+ B cell   \n",
       "\n",
       "                    predicted.celltype.l3  predicted.celltype.l4  \\\n",
       "AAACCTGCAACGATCT-1                B naive                 B cell   \n",
       "AAACGGGCAGTTCATG-1                B naive                 B cell   \n",
       "AAACGGGGTCCCTACT-1                B naive                 B cell   \n",
       "AAACGGGGTTGCGCAC-1                B naive                 B cell   \n",
       "AAAGATGGTTATGCGT-1                B naive                 B cell   \n",
       "\n",
       "                   predicted.celltype.l5 predicted.celltype.l6 G2M.Score  \\\n",
       "AAACCTGCAACGATCT-1                B cell              Lymphoid -0.026746   \n",
       "AAACGGGCAGTTCATG-1                B cell              Lymphoid -0.022354   \n",
       "AAACGGGGTCCCTACT-1                B cell              Lymphoid  0.013420   \n",
       "AAACGGGGTTGCGCAC-1                B cell              Lymphoid  0.006212   \n",
       "AAAGATGGTTATGCGT-1                B cell              Lymphoid -0.011067   \n",
       "\n",
       "                     S.Score Phase             barcode  \n",
       "AAACCTGCAACGATCT-1 -0.000268    G1  AAACCTGCAACGATCT-1  \n",
       "AAACGGGCAGTTCATG-1  0.018777     S  AAACGGGCAGTTCATG-1  \n",
       "AAACGGGGTCCCTACT-1 -0.004957   G2M  AAACGGGGTCCCTACT-1  \n",
       "AAACGGGGTTGCGCAC-1  0.010032     S  AAACGGGGTTGCGCAC-1  \n",
       "AAAGATGGTTATGCGT-1  0.004024     S  AAAGATGGTTATGCGT-1  \n",
       "\n",
       "[5 rows x 23 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cov_naive_Bcell.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1282ff18",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "79cbfda7",
   "metadata": {},
   "outputs": [],
   "source": [
    "phenotype = phenotype.sel(cell=cov_naive_Bcell[\"barcode\"].values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "32e3a827",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><svg style=\"position: absolute; width: 0; height: 0; overflow: hidden\">\n",
       "<defs>\n",
       "<symbol id=\"icon-database\" viewBox=\"0 0 32 32\">\n",
       "<path d=\"M16 0c-8.837 0-16 2.239-16 5v4c0 2.761 7.163 5 16 5s16-2.239 16-5v-4c0-2.761-7.163-5-16-5z\"></path>\n",
       "<path d=\"M16 17c-8.837 0-16-2.239-16-5v6c0 2.761 7.163 5 16 5s16-2.239 16-5v-6c0 2.761-7.163 5-16 5z\"></path>\n",
       "<path d=\"M16 26c-8.837 0-16-2.239-16-5v6c0 2.761 7.163 5 16 5s16-2.239 16-5v-6c0 2.761-7.163 5-16 5z\"></path>\n",
       "</symbol>\n",
       "<symbol id=\"icon-file-text2\" viewBox=\"0 0 32 32\">\n",
       "<path d=\"M28.681 7.159c-0.694-0.947-1.662-2.053-2.724-3.116s-2.169-2.030-3.116-2.724c-1.612-1.182-2.393-1.319-2.841-1.319h-15.5c-1.378 0-2.5 1.121-2.5 2.5v27c0 1.378 1.122 2.5 2.5 2.5h23c1.378 0 2.5-1.122 2.5-2.5v-19.5c0-0.448-0.137-1.23-1.319-2.841zM24.543 5.457c0.959 0.959 1.712 1.825 2.268 2.543h-4.811v-4.811c0.718 0.556 1.584 1.309 2.543 2.268zM28 29.5c0 0.271-0.229 0.5-0.5 0.5h-23c-0.271 0-0.5-0.229-0.5-0.5v-27c0-0.271 0.229-0.5 0.5-0.5 0 0 15.499-0 15.5 0v7c0 0.552 0.448 1 1 1h7v19.5z\"></path>\n",
       "<path d=\"M23 26h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z\"></path>\n",
       "<path d=\"M23 22h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z\"></path>\n",
       "<path d=\"M23 18h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z\"></path>\n",
       "</symbol>\n",
       "</defs>\n",
       "</svg>\n",
       "<style>/* CSS stylesheet for displaying xarray objects in jupyterlab.\n",
       " *\n",
       " */\n",
       "\n",
       ":root {\n",
       "  --xr-font-color0: var(--jp-content-font-color0, rgba(0, 0, 0, 1));\n",
       "  --xr-font-color2: var(--jp-content-font-color2, rgba(0, 0, 0, 0.54));\n",
       "  --xr-font-color3: var(--jp-content-font-color3, rgba(0, 0, 0, 0.38));\n",
       "  --xr-border-color: var(--jp-border-color2, #e0e0e0);\n",
       "  --xr-disabled-color: var(--jp-layout-color3, #bdbdbd);\n",
       "  --xr-background-color: var(--jp-layout-color0, white);\n",
       "  --xr-background-color-row-even: var(--jp-layout-color1, white);\n",
       "  --xr-background-color-row-odd: var(--jp-layout-color2, #eeeeee);\n",
       "}\n",
       "\n",
       "html[theme=dark],\n",
       "body.vscode-dark {\n",
       "  --xr-font-color0: rgba(255, 255, 255, 1);\n",
       "  --xr-font-color2: rgba(255, 255, 255, 0.54);\n",
       "  --xr-font-color3: rgba(255, 255, 255, 0.38);\n",
       "  --xr-border-color: #1F1F1F;\n",
       "  --xr-disabled-color: #515151;\n",
       "  --xr-background-color: #111111;\n",
       "  --xr-background-color-row-even: #111111;\n",
       "  --xr-background-color-row-odd: #313131;\n",
       "}\n",
       "\n",
       ".xr-wrap {\n",
       "  display: block !important;\n",
       "  min-width: 300px;\n",
       "  max-width: 700px;\n",
       "}\n",
       "\n",
       ".xr-text-repr-fallback {\n",
       "  /* fallback to plain text repr when CSS is not injected (untrusted notebook) */\n",
       "  display: none;\n",
       "}\n",
       "\n",
       ".xr-header {\n",
       "  padding-top: 6px;\n",
       "  padding-bottom: 6px;\n",
       "  margin-bottom: 4px;\n",
       "  border-bottom: solid 1px var(--xr-border-color);\n",
       "}\n",
       "\n",
       ".xr-header > div,\n",
       ".xr-header > ul {\n",
       "  display: inline;\n",
       "  margin-top: 0;\n",
       "  margin-bottom: 0;\n",
       "}\n",
       "\n",
       ".xr-obj-type,\n",
       ".xr-array-name {\n",
       "  margin-left: 2px;\n",
       "  margin-right: 10px;\n",
       "}\n",
       "\n",
       ".xr-obj-type {\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-sections {\n",
       "  padding-left: 0 !important;\n",
       "  display: grid;\n",
       "  grid-template-columns: 150px auto auto 1fr 20px 20px;\n",
       "}\n",
       "\n",
       ".xr-section-item {\n",
       "  display: contents;\n",
       "}\n",
       "\n",
       ".xr-section-item input {\n",
       "  display: none;\n",
       "}\n",
       "\n",
       ".xr-section-item input + label {\n",
       "  color: var(--xr-disabled-color);\n",
       "}\n",
       "\n",
       ".xr-section-item input:enabled + label {\n",
       "  cursor: pointer;\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-section-item input:enabled + label:hover {\n",
       "  color: var(--xr-font-color0);\n",
       "}\n",
       "\n",
       ".xr-section-summary {\n",
       "  grid-column: 1;\n",
       "  color: var(--xr-font-color2);\n",
       "  font-weight: 500;\n",
       "}\n",
       "\n",
       ".xr-section-summary > span {\n",
       "  display: inline-block;\n",
       "  padding-left: 0.5em;\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:disabled + label {\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-section-summary-in + label:before {\n",
       "  display: inline-block;\n",
       "  content: '►';\n",
       "  font-size: 11px;\n",
       "  width: 15px;\n",
       "  text-align: center;\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:disabled + label:before {\n",
       "  color: var(--xr-disabled-color);\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:checked + label:before {\n",
       "  content: '▼';\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:checked + label > span {\n",
       "  display: none;\n",
       "}\n",
       "\n",
       ".xr-section-summary,\n",
       ".xr-section-inline-details {\n",
       "  padding-top: 4px;\n",
       "  padding-bottom: 4px;\n",
       "}\n",
       "\n",
       ".xr-section-inline-details {\n",
       "  grid-column: 2 / -1;\n",
       "}\n",
       "\n",
       ".xr-section-details {\n",
       "  display: none;\n",
       "  grid-column: 1 / -1;\n",
       "  margin-bottom: 5px;\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:checked ~ .xr-section-details {\n",
       "  display: contents;\n",
       "}\n",
       "\n",
       ".xr-array-wrap {\n",
       "  grid-column: 1 / -1;\n",
       "  display: grid;\n",
       "  grid-template-columns: 20px auto;\n",
       "}\n",
       "\n",
       ".xr-array-wrap > label {\n",
       "  grid-column: 1;\n",
       "  vertical-align: top;\n",
       "}\n",
       "\n",
       ".xr-preview {\n",
       "  color: var(--xr-font-color3);\n",
       "}\n",
       "\n",
       ".xr-array-preview,\n",
       ".xr-array-data {\n",
       "  padding: 0 5px !important;\n",
       "  grid-column: 2;\n",
       "}\n",
       "\n",
       ".xr-array-data,\n",
       ".xr-array-in:checked ~ .xr-array-preview {\n",
       "  display: none;\n",
       "}\n",
       "\n",
       ".xr-array-in:checked ~ .xr-array-data,\n",
       ".xr-array-preview {\n",
       "  display: inline-block;\n",
       "}\n",
       "\n",
       ".xr-dim-list {\n",
       "  display: inline-block !important;\n",
       "  list-style: none;\n",
       "  padding: 0 !important;\n",
       "  margin: 0;\n",
       "}\n",
       "\n",
       ".xr-dim-list li {\n",
       "  display: inline-block;\n",
       "  padding: 0;\n",
       "  margin: 0;\n",
       "}\n",
       "\n",
       ".xr-dim-list:before {\n",
       "  content: '(';\n",
       "}\n",
       "\n",
       ".xr-dim-list:after {\n",
       "  content: ')';\n",
       "}\n",
       "\n",
       ".xr-dim-list li:not(:last-child):after {\n",
       "  content: ',';\n",
       "  padding-right: 5px;\n",
       "}\n",
       "\n",
       ".xr-has-index {\n",
       "  font-weight: bold;\n",
       "}\n",
       "\n",
       ".xr-var-list,\n",
       ".xr-var-item {\n",
       "  display: contents;\n",
       "}\n",
       "\n",
       ".xr-var-item > div,\n",
       ".xr-var-item label,\n",
       ".xr-var-item > .xr-var-name span {\n",
       "  background-color: var(--xr-background-color-row-even);\n",
       "  margin-bottom: 0;\n",
       "}\n",
       "\n",
       ".xr-var-item > .xr-var-name:hover span {\n",
       "  padding-right: 5px;\n",
       "}\n",
       "\n",
       ".xr-var-list > li:nth-child(odd) > div,\n",
       ".xr-var-list > li:nth-child(odd) > label,\n",
       ".xr-var-list > li:nth-child(odd) > .xr-var-name span {\n",
       "  background-color: var(--xr-background-color-row-odd);\n",
       "}\n",
       "\n",
       ".xr-var-name {\n",
       "  grid-column: 1;\n",
       "}\n",
       "\n",
       ".xr-var-dims {\n",
       "  grid-column: 2;\n",
       "}\n",
       "\n",
       ".xr-var-dtype {\n",
       "  grid-column: 3;\n",
       "  text-align: right;\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-var-preview {\n",
       "  grid-column: 4;\n",
       "}\n",
       "\n",
       ".xr-var-name,\n",
       ".xr-var-dims,\n",
       ".xr-var-dtype,\n",
       ".xr-preview,\n",
       ".xr-attrs dt {\n",
       "  white-space: nowrap;\n",
       "  overflow: hidden;\n",
       "  text-overflow: ellipsis;\n",
       "  padding-right: 10px;\n",
       "}\n",
       "\n",
       ".xr-var-name:hover,\n",
       ".xr-var-dims:hover,\n",
       ".xr-var-dtype:hover,\n",
       ".xr-attrs dt:hover {\n",
       "  overflow: visible;\n",
       "  width: auto;\n",
       "  z-index: 1;\n",
       "}\n",
       "\n",
       ".xr-var-attrs,\n",
       ".xr-var-data {\n",
       "  display: none;\n",
       "  background-color: var(--xr-background-color) !important;\n",
       "  padding-bottom: 5px !important;\n",
       "}\n",
       "\n",
       ".xr-var-attrs-in:checked ~ .xr-var-attrs,\n",
       ".xr-var-data-in:checked ~ .xr-var-data {\n",
       "  display: block;\n",
       "}\n",
       "\n",
       ".xr-var-data > table {\n",
       "  float: right;\n",
       "}\n",
       "\n",
       ".xr-var-name span,\n",
       ".xr-var-data,\n",
       ".xr-attrs {\n",
       "  padding-left: 25px !important;\n",
       "}\n",
       "\n",
       ".xr-attrs,\n",
       ".xr-var-attrs,\n",
       ".xr-var-data {\n",
       "  grid-column: 1 / -1;\n",
       "}\n",
       "\n",
       "dl.xr-attrs {\n",
       "  padding: 0;\n",
       "  margin: 0;\n",
       "  display: grid;\n",
       "  grid-template-columns: 125px auto;\n",
       "}\n",
       "\n",
       ".xr-attrs dt,\n",
       ".xr-attrs dd {\n",
       "  padding: 0;\n",
       "  margin: 0;\n",
       "  float: left;\n",
       "  padding-right: 10px;\n",
       "  width: auto;\n",
       "}\n",
       "\n",
       ".xr-attrs dt {\n",
       "  font-weight: normal;\n",
       "  grid-column: 1;\n",
       "}\n",
       "\n",
       ".xr-attrs dt:hover span {\n",
       "  display: inline-block;\n",
       "  background: var(--xr-background-color);\n",
       "  padding-right: 10px;\n",
       "}\n",
       "\n",
       ".xr-attrs dd {\n",
       "  grid-column: 2;\n",
       "  white-space: pre-wrap;\n",
       "  word-break: break-all;\n",
       "}\n",
       "\n",
       ".xr-icon-database,\n",
       ".xr-icon-file-text2 {\n",
       "  display: inline-block;\n",
       "  vertical-align: middle;\n",
       "  width: 1em;\n",
       "  height: 1.5em !important;\n",
       "  stroke-width: 0;\n",
       "  stroke: currentColor;\n",
       "  fill: currentColor;\n",
       "}\n",
       "</style><pre class='xr-text-repr-fallback'>&lt;xarray.DataArray (trait: 600, cell: 71374)&gt;\n",
       "array([[0., 0., 0., ..., 0., 0., 0.],\n",
       "       [0., 0., 0., ..., 0., 0., 0.],\n",
       "       [0., 0., 0., ..., 0., 0., 0.],\n",
       "       ...,\n",
       "       [0., 0., 0., ..., 0., 0., 0.],\n",
       "       [0., 0., 0., ..., 0., 0., 0.],\n",
       "       [0., 0., 0., ..., 0., 0., 0.]])\n",
       "Coordinates:\n",
       "  * trait    (trait) object &#x27;AC002472.1&#x27; &#x27;KIAA1658&#x27; ... &#x27;AC000036.4&#x27;\n",
       "  * cell     (cell) object &#x27;AAACCTGCAACGATCT-1&#x27; ... &#x27;TTTGTCAGTGTGAAAT-9&#x27;</pre><div class='xr-wrap' style='display:none'><div class='xr-header'><div class='xr-obj-type'>xarray.DataArray</div><div class='xr-array-name'></div><ul class='xr-dim-list'><li><span class='xr-has-index'>trait</span>: 600</li><li><span class='xr-has-index'>cell</span>: 71374</li></ul></div><ul class='xr-sections'><li class='xr-section-item'><div class='xr-array-wrap'><input id='section-4addb32b-b8e5-4457-b665-fa0294d5c350' class='xr-array-in' type='checkbox' checked><label for='section-4addb32b-b8e5-4457-b665-fa0294d5c350' title='Show/hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-array-preview xr-preview'><span>0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 ... 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0</span></div><div class='xr-array-data'><pre>array([[0., 0., 0., ..., 0., 0., 0.],\n",
       "       [0., 0., 0., ..., 0., 0., 0.],\n",
       "       [0., 0., 0., ..., 0., 0., 0.],\n",
       "       ...,\n",
       "       [0., 0., 0., ..., 0., 0., 0.],\n",
       "       [0., 0., 0., ..., 0., 0., 0.],\n",
       "       [0., 0., 0., ..., 0., 0., 0.]])</pre></div></div></li><li class='xr-section-item'><input id='section-4bcdfdcb-5566-4b53-bac6-ea53a81ecb4c' class='xr-section-summary-in' type='checkbox'  checked><label for='section-4bcdfdcb-5566-4b53-bac6-ea53a81ecb4c' class='xr-section-summary' >Coordinates: <span>(2)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><ul class='xr-var-list'><li class='xr-var-item'><div class='xr-var-name'><span class='xr-has-index'>trait</span></div><div class='xr-var-dims'>(trait)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;AC002472.1&#x27; ... &#x27;AC000036.4&#x27;</div><input id='attrs-adc375df-3af5-4bb6-983f-05e930c102ef' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-adc375df-3af5-4bb6-983f-05e930c102ef' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-c14ce6b8-5098-4ffa-84f0-a23a55aa94d6' class='xr-var-data-in' type='checkbox'><label for='data-c14ce6b8-5098-4ffa-84f0-a23a55aa94d6' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;AC002472.1&#x27;, &#x27;KIAA1658&#x27;, &#x27;AC005003.1&#x27;, ..., &#x27;CHKB-AS1&#x27;, &#x27;CTA-384D8.33&#x27;,\n",
       "       &#x27;AC000036.4&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span class='xr-has-index'>cell</span></div><div class='xr-var-dims'>(cell)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;AAACCTGCAACGATCT-1&#x27; ... &#x27;TTTGTC...</div><input id='attrs-beece3dd-60e7-48a7-bc37-4e95a6d9f486' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-beece3dd-60e7-48a7-bc37-4e95a6d9f486' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-dc297431-a27e-4833-a0dc-c0d08362d885' class='xr-var-data-in' type='checkbox'><label for='data-dc297431-a27e-4833-a0dc-c0d08362d885' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;AAACCTGCAACGATCT-1&#x27;, &#x27;AAACGGGCAGTTCATG-1&#x27;, &#x27;AAACGGGGTCCCTACT-1&#x27;, ...,\n",
       "       &#x27;TTTGGTTTCGGCGCTA-9&#x27;, &#x27;TTTGTCAGTCTAGCGC-9&#x27;, &#x27;TTTGTCAGTGTGAAAT-9&#x27;],\n",
       "      dtype=object)</pre></div></li></ul></div></li><li class='xr-section-item'><input id='section-b14399f9-dcb6-40df-b6b5-ed1816786415' class='xr-section-summary-in' type='checkbox' disabled ><label for='section-b14399f9-dcb6-40df-b6b5-ed1816786415' class='xr-section-summary'  title='Expand/collapse section'>Attributes: <span>(0)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><dl class='xr-attrs'></dl></div></li></ul></div></div>"
      ],
      "text/plain": [
       "<xarray.DataArray (trait: 600, cell: 71374)>\n",
       "array([[0., 0., 0., ..., 0., 0., 0.],\n",
       "       [0., 0., 0., ..., 0., 0., 0.],\n",
       "       [0., 0., 0., ..., 0., 0., 0.],\n",
       "       ...,\n",
       "       [0., 0., 0., ..., 0., 0., 0.],\n",
       "       [0., 0., 0., ..., 0., 0., 0.],\n",
       "       [0., 0., 0., ..., 0., 0., 0.]])\n",
       "Coordinates:\n",
       "  * trait    (trait) object 'AC002472.1' 'KIAA1658' ... 'AC000036.4'\n",
       "  * cell     (cell) object 'AAACCTGCAACGATCT-1' ... 'TTTGTCAGTGTGAAAT-9'"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# still 600 genes, but \"only\" ~70k cells\n",
    "phenotype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5d602656",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "327eaedd",
   "metadata": {},
   "outputs": [],
   "source": [
    "## We want to indentify outlier donors\n",
    "## Strategy 1:\n",
    "### take sum / mean / median / variance across cells per donor\n",
    "### split AGGREGATED expression of genes in quantiles (Z-scores?), consider outlier values=donors\n",
    "\n",
    "## Strategy 2:\n",
    "### split SINGLE_CELL expression of genes in quantiles (Z-scores?), consider outlier values\n",
    "### divide those counts but total cells from that individual"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a2c1ae35",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Marc's paper outlier detection procedure\n",
    "# * Based on featureCount gene quantifications (log TPM), we considered autosomal protein-coding and long noncoding RNA genes. \n",
    "# * Cell lines from donors with predicted ancestry other than the European super-population were discarded\n",
    "# * and we additionally limited the analysis to lines with paired-end RNA-seq data\n",
    "# * Genes were filtered for minimal expression, defined as gene expression TPM > 0 in 50% or more in each study.\n",
    "# * To adjust for transcriptome-wide confounders, PEER-correction was run on the filtered data (n = 50 PEER factors). \n",
    "# * The resulting residual expression profiles were scaled and centered (z-score normalization). \n",
    "# * cell lines with an expression absolute z-score > 2 in > 100 genes were discarded from subsequent analyses. \n",
    "# * Finally, cell lines were retained if WGS data were available in addition to RNA-seq data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "58e3c05a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# * variants were filtered based on the variant quality score recalibration method using a tranche cutoff of 99%. \n",
    "# * The software vcfanno (v0.2.9) was used to annotate the WGS vcf file with MAF from gnomAD (v2.0.2) and \n",
    "# CADD score from CADD (v1.3). \n",
    "# * Variants were filtered on a per-sample level to retain variants with at least one alternate allele. \n",
    "# * Variants were then linked to genes using the bcftools (v1.11) window command, selecting a maximum distance of\n",
    "# 10 kb based on the Ensembl 75 GTF reference. \n",
    "# * A separate file was produced for each cell line consisting of the following columns: \n",
    "# [cell line ID; gene ID; chromosome; position; gnomAD MAF; CADD (phred); and CADD (raw).]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "28255cb7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# * We considered the subset of lines with both RNA-seq and WGS data available \n",
    "# * and focused on variants up to 10 kb upstream and downstream of the protein-coding and long noncoding RNA genes. \n",
    "# * Gene expression outliers for a given gene were defined as samples with a minimum gene expression z-score \n",
    "# (z-score < −2; underexpression outlier) or a maximum gene expression z-score (z-score > 2; overexpression outlier). \n",
    "# * Separate scores were computed for gene-level underexpression outliers and overexpression outliers. \n",
    "# * The reported enrichment score was calculated as the ratio of the proportion of outlier lines with variants across \n",
    "# several MAF/CADD bins compared to non-outlier lines. \n",
    "# Specifically, enrichment here refers to the relative risk (RR):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "03359a2a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# * The analysis was performed separately for SNPs, indels and SVs, and across different \n",
    "# MAF bins (from common to rare) and CADD bins (progressively more deleterious variants). \n",
    "# * Genes were discarded if there was not at least 1 outlier and 1 non-outlier line matching MAF and CADD thresholds"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (Jupyter - cellregmap)",
   "language": "python",
   "name": "cellregmap_notebook"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
